### PR TITLE
Basic test framework to make testing cohttp easier

### PR DESCRIPTION
**Pull request by rgrinberg**
Cohttp needs a lot more end to end testing so this is a crude attempt at
making it easier.

Suggestions on how to handle async as well are welcome

@avsm I've whined enough about this that it's time to address this problem

### CODE REVIEW

**Comment by rgrinberg** [comment]
bump

**Comment by avsm** [comment]
Bit baffled by the Travis failure:

```
[31m#===[m [01mERROR[m [01mwhile installing cstruct.1.5.0[m [31m=====================================#[m
[33m#[m [33mopam-version[m 1.2.1
[33m#[m [33mos[m           linux
[33m#[m [33mcommand[m      make test
[33m#[m [33mpath[m         /home/travis/.opam/system/build/cstruct.1.5.0
[33m#[m [33mcompiler[m     system (4.02.1)
[33m#[m [33mexit-code[m    2
[33m#[m [33menv-file[m     /home/travis/.opam/system/build/cstruct.1.5.0/cstruct-1231-0ac68e.env
[33m#[m [33mstdout-file[m  /home/travis/.opam/system/build/cstruct.1.5.0/cstruct-1231-0ac68e.out
[33m#[m [33mstderr-file[m  /home/travis/.opam/system/build/cstruct.1.5.0/cstruct-1231-0ac68e.err
[33m### stdout ###
[m[33m# [m[...]
[33m# [mWarning: the tag "tests" is not used in any flag declaration, so it will have no effect; it may be a typo. Otherwise use `mark_tag_used` in your myocamlbuild.ml to disable this warning.
[33m# [m/home/travis/.opam/system/bin/ocamlfind ocamlopt -shared -I lib/ lib/cstruct.cmxa lib/cstruct.cmx -o lib/cstruct.cmxs
[33m# [m/home/travis/.opam/system/bin/ocamlfind ocamlopt -shared async/async_cstruct.cmxa async/async_cstruct.cmx -o async/async_cstruct.cmxs
[33m# [m/home/travis/.opam/system/bin/ocamlfind ocamlopt -shared lwt/lwt_cstruct.cmxa lwt/lwt_cstruct.cmx -o lwt/lwt_cstruct.cmxs
[33m# [m/home/travis/.opam/system/bin/ocamlfind ocamlopt -shared unix/unix_cstruct.cmxa unix/unix_cstruct.cmx -o unix/unix_cstruct.cmxs
[33m# [m/home/travis/.opam/system/bin/ocamlfind ocamlopt -shared syntax/cstruct-syntax.cmxa syntax/pa_cstruct.cmx -o syntax/cstruct-syntax.cmxs
[33m# [m/home/travis/.opam/system/bin/ocamlfind ocamldep -package bigarray -package oUnit -package ocplib-endian -package ocplib-endian.bigstring -package sexplib -syntax camlp4o -package sexplib.syntax -package unix -modules lib_test/bounds.ml > lib_test/bounds.ml.depends
[33m# [m/home/travis/.opam/system/bin/ocamlfind ocamlc -c -g -annot -bin-annot -I lib -package bigarray -package oUnit -package ocplib-endian -package ocplib-endian.bigstring -package sexplib -syntax camlp4o -package sexplib.syntax -package unix -I lib_test -I lib -I async -I lwt -I unix -o lib_test/bounds.cmo lib_test/bounds.ml
[33m# [m/home/travis/.opam/system/bin/ocamlfind ocamlc -g -I lib/ -linkpkg -package bigarray -package oUnit -package ocplib-endian -package ocplib-endian.bigstring -package sexplib -package sexplib.syntax -package unix lib/cstruct.cma lib_test/bounds.cmo -o lib_test/bounds.byte
[33m# [mocaml setup.ml -test 
[33m### stderr ###
[m[33m# [mFatal error: cannot load shared library dllcstruct_stubs
[33m# [mReason: dllcstruct_stubs.so: cannot open shared object file: No such file or directory
[33m# [mW: Test 'test_bounds' fails: Command '/home/travis/.opam/system/build/cstruct.1.5.0/_build/lib_test/bounds.byte' terminated with error code 2
[33m# [mE: Failure("Tests had a 100.00% failure rate")
[33m# [mmake: *** [test] Error 1
```

**Comment by rgrinberg** [comment]
Aye, that one has been failing for a while now. Although I don't remember that particular error.

File: lwt/cohttp_lwt_unix_test.ml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      x
+    | [] -> Lwt.fail_with "response_sequence: Server exhausted responses"
+
+let temp_server ?port spec callback =
+  let port = match port with
+    | None -> next_port ()
+    | Some p -> p in
+  let server = Server.make ~callback:(fun _ req body -> spec req body) () in
+  let uri = Uri.of_string ("http://0.0.0.0:" ^ (string_of_int port)) in
+  let server_ready = Lwt_unix.sleep 0.3 in (* XXX hackish *)
```
Comment (avsm) [inline lwt/cohttp_lwt_unix_test.ml]: Are you finding that there's a race between `Server.create` and connections being accepted?  It ought to block until the socket is open and listening.

Comment (rgrinberg) [inline lwt/cohttp_lwt_unix_test.ml]: Oops I forgot to get rid of this.

**Comment by rgrinberg** [comment]
@dsheets @avsm I'm merging this because It's annoying to keep rebasing. This PR doesn't touch any of the library code and doesn't expose anything to users. If there are any objections there will be no risk resolving them later.

**Pull request merged**